addons/ai_apps/models/discuss_channel.py

```diff
@@ -0,0 +1,208 @@
+            temp_context.append({
+                'role': 'assistant',
+                'content': 'Hello, how can I rewrite your text?',
+            })
+
+        self.ai_context = temp_context
+
+    @api.model
+    def create_ai_composer_channel(self, caller_component, record_name, record_model=None, record_id=None, textSelection=None):
+        # create a new AI chat
+        channel = self.create({
+            'channel_member_ids': [
+                Command.create({
+                    'partner_id': self.env.user.partner_id.id,
+                }),
+            ],
+            'channel_type': 'ai_composer',
+            'name': 'AI: ' + record_name,
+        })
+
+        # fetch the record's info and feed it into context
+        channel._initialise_context(record_model, record_id, caller_component, textSelection)
+
+        return channel
+
+    def add_message_to_context(self, message, author):
+        # here I tried doing self.ai_context.append(...) but it didn't work :()
+        current_context = self.ai_context
+        current_context.append({
+            'role': author,
+            'content': message,
+        })
+        self.ai_context = current_context
+
+    def submit_to_model(self, prompt, conversation_history):
+        try:
+            IrConfigParameter = self.env['ir.config_parameter'].sudo()
+            olg_api_endpoint = IrConfigParameter.get_param('web_editor.olg_api_endpoint', DEFAULT_OLG_ENDPOINT)
+            database_id = IrConfigParameter.get_param('database.uuid')
+            print(conversation_history)
+            response = iap_tools.iap_jsonrpc(olg_api_endpoint + "/api/olg/1/chat", params={
+                'prompt': prompt,
+                'conversation_history': conversation_history or [],
+                'database_id': database_id,
+            }, timeout=30)
+            if response['status'] == 'success':
+                return response['content']
+            elif response['status'] == 'error_prompt_too_long':
+                return "⚠️ Sorry, your prompt is too long. Try to say it in fewer words."
+            elif response['status'] == 'limit_call_reached':
+                return "⚠️ You have reached the maximum number of requests for this service. Try again later."
+            else:
+                return "⚠️ Sorry, we could not generate a response. Please try again later."
+        except AccessError:
+            return "⚠️ Oops, it looks like our AI is unreachable!"
+
+    def create_response(self, model_response):
+        # add the model response in the context
+        self.add_message_to_context(model_response, 'assistant')
+        # translate the markdown formatted text into HTML and sanitize the HTML
+        response_html = markdown(model_response, extensions=["fenced_code", "tables"])
+        final_response = Markup(html_sanitize(response_html))
+        odoobot_id = self.env['ir.model.data']._xmlid_to_res_id("base.partner_root")
+        self.sudo().message_post(
```

Comment:
We're trying to add small `# sudo:` comments to explain why it is necessary and acceptable to do so when using it

Reply:
It seems like we don't need it after all, removed it and everything works as it should :shrug: